File: Python/ml_methods/svd/demo_svd_01_in_memory.py
# ==============================================================================
# demo_svd_01_in_memory.py
# ==============================================================================
#
# Demo illustrating SVDSolver usage on in-memory data.
# Mirrors cpp/ml_methods/svd/demo_mlm_svd_01/demo_mlm_svd_01.cpp.
#
# Demo content:
#   - Part A: Direct / thin-Jacobi path (n = 300, p = 80, M = 10) — truncated
#             SVD, relative reconstruction error, top singular values, plus a
#             full-rank (M = p) sanity check
#   - Part B: Gram / kernel path (n = 100, p = 500, M = 8) — wide matrices
#             with p > 2*n route through the n x n Gram matrix
#   - Part C: Orthogonality check (n = 200, p = 150, M = 20) — U and V must
#             have orthonormal columns
#
# Notes:
#   - SVDSolver.compute(X, M) returns the rank-M truncated SVD: it keeps only
#     the M largest singular values (S), the corresponding left singular
#     vectors (U, n x M) and right singular vectors (V, p x M), so
#     U @ diag(S) @ V.T is the best rank-M approximation of X in the
#     Frobenius norm (Eckart-Young).
#   - The C++ demo seeds std::mt19937; here we seed numpy's default_rng with
#     the same values. The RNG streams differ, so the printed numbers match
#     the C++ output qualitatively, not bitwise.
#
# ==============================================================================

import numpy as np
from trex_selector_neo.ml_methods import SVDSolver

print("\n" + "=" * 70)
print("SVDSolver Demos (in-memory)")
print("=" * 70 + "\n")


# ==============================================================================
# Part A: Direct / Thin-Jacobi Path (n >= p)
# ==============================================================================

print("-" * 70)
print("Part A: SVDSolver - Direct Path (n >= p)")
print("-" * 70 + "\n")

n, p, M = 300, 80, 10
print(f"Matrix size : {n} x {p}")
print(f"M           : {M}\n")

rng = np.random.default_rng(42)
# The solver accesses the matrix through Eigen and therefore requires a
# Fortran-ordered (column-major) float64 array.
X = np.asfortranarray(rng.standard_normal((n, p)))

solver = SVDSolver()
res = solver.compute(X, M)

print(f"U dimensions : {res.U.shape[0]} x {res.U.shape[1]}")
print(f"S size       : {res.S.size}")
print(f"V dimensions : {res.V.shape[0]} x {res.V.shape[1]}\n")

# Relative reconstruction error ||X_M - X||_F / ||X||_F. With only M = 10 of
# 80 possible components of a random Gaussian matrix, most of the energy is
# discarded, so this error is large by design — truncation keeps only the
# dominant directions.
X_approx = res.U @ np.diag(res.S) @ res.V.T
rel_error = np.linalg.norm(X_approx - X) / np.linalg.norm(X)
print(f"Relative reconstruction error ||X_M - X||_F / ||X||_F : {rel_error:.6e}")

# Singular values come back sorted in decreasing order.
print(f"\nTop {M} singular values:")
for k in range(M):
    print(f"  s{k + 1} = {res.S[k]:.4f}")
assert np.all(np.diff(res.S) <= 1e-12), "singular values must be decreasing"

# Didactic check: at full rank (M = p) the truncated SVD reproduces X exactly
# up to floating-point error.
res_full = solver.compute(X, p)
X_full = res_full.U @ np.diag(res_full.S) @ res_full.V.T
rel_error_full = np.linalg.norm(X_full - X) / np.linalg.norm(X)
print(f"\nFull-rank (M = {p}) reconstruction error: {rel_error_full:.6e}")
assert rel_error_full < 1e-10, "full-rank SVD must reconstruct X exactly"
print("Full-rank reconstruction check passed")


# ==============================================================================
# Part B: Gram / Kernel Path (p > 2*n)
# ==============================================================================

print("\n" + "-" * 70)
print("Part B: SVDSolver - Gram Path (p > 2*n)")
print("-" * 70 + "\n")

n, p, M = 100, 500, 8
print(f"Matrix size : {n} x {p}  (p > 2*n -> Gram path)")
print(f"M           : {M}\n")

rng = np.random.default_rng(7)
X = np.asfortranarray(rng.standard_normal((n, p)))

# For very wide matrices the solver works with the n x n Gram matrix X X^T
# instead of X itself — the same truncated factorization at far lower cost.
solver = SVDSolver()
res = solver.compute(X, M)

print(f"U dimensions : {res.U.shape[0]} x {res.U.shape[1]}")
print(f"S size       : {res.S.size}")
print(f"V dimensions : {res.V.shape[0]} x {res.V.shape[1]}\n")

X_approx = res.U @ np.diag(res.S) @ res.V.T
rel_error = np.linalg.norm(X_approx - X) / np.linalg.norm(X)
print(f"Relative reconstruction error ||X_M - X||_F / ||X||_F : {rel_error:.6e}")


# ==============================================================================
# Part C: Orthogonality Check
# ==============================================================================

print("\n" + "-" * 70)
print("Part C: SVDSolver - Orthogonality Check")
print("-" * 70 + "\n")

n, p, M = 200, 150, 20
print(f"Matrix size : {n} x {p}")
print(f"M           : {M}\n")

rng = np.random.default_rng(99)
X = np.asfortranarray(rng.standard_normal((n, p)))

solver = SVDSolver()
res = solver.compute(X, M)

# Singular vectors are orthonormal: U^T U = I and V^T V = I (M x M identity).
UtU_error = np.linalg.norm(res.U.T @ res.U - np.eye(M))
VtV_error = np.linalg.norm(res.V.T @ res.V - np.eye(M))

print(f"||U^T U - I||_F : {UtU_error:.6e}")
print(f"||V^T V - I||_F : {VtV_error:.6e}\n")

assert UtU_error < 1e-10 and VtV_error < 1e-10, "orthonormality check failed"
print("U and V are orthonormal")

print("\n" + "=" * 70)
print("demo_svd_01_in_memory.py complete.")

File: R/ml_methods/svd/demo_svd_01_in_memory.R
# ==============================================================================
# demo_svd_01_in_memory.R
# ==============================================================================
#
# Demo illustrating SVDSolver (truncated rank-M SVD) on in-memory data.
#
# Demo content:
#   - Part 1: Direct path (n >= p) — truncated and full-rank reconstruction
#   - Part 2: Gram / kernel path (p > 2*n) — wide-matrix decomposition
#   - Part 3: Orthogonality check — ||U^T U - I||_F and ||V^T V - I||_F
#
# Mirrors: cpp/ml_methods/svd/demo_mlm_svd_01/demo_mlm_svd_01.cpp
#
# Note: same n / p / M values and seeds as the C++ demo; R's Mersenne-Twister
# stream differs from std::mt19937, so results match qualitatively, not
# bitwise.
#
# ==============================================================================

library(TRexSelectorNeo)

# ==============================================================================

cat("\n", strrep("=", 70), "\n", sep = "")
cat("SVDSolver Demos (in-memory)\n")
cat(strrep("=", 70), "\n\n")

# The truncated SVD keeps the M leading singular triplets of X (n x p):
#   X_M = U_M * diag(S_M) * V_M^T,   U_M: n x M,  S_M: M,  V_M: p x M
# X_M is the best rank-M approximation of X in the Frobenius norm
# (Eckart-Young). $compute(X, M) returns a list with elements U, S, V.

solver <- SVDSolver$new()

# ==============================================================================
# Part 1: Direct / Thin Path (n >= p)
# ==============================================================================
# For tall matrices the solver decomposes X directly.
# ==============================================================================

cat(strrep("-", 70), "\n")
cat("Part 1: Direct Path (n >= p)\n")
cat(strrep("-", 70), "\n\n")

set.seed(42)
n1 <- 300L
p1 <- 80L
M1 <- 10L

cat(sprintf("Matrix size : %d x %d\n", n1, p1))
cat(sprintf("M           : %d\n\n", M1))

X1 <- matrix(rnorm(n1 * p1), nrow = n1, ncol = p1)

res1 <- solver$compute(X1, M1)

cat(sprintf("U dimensions : %d x %d\n", nrow(res1$U), ncol(res1$U)))
cat(sprintf("S length     : %d\n", length(res1$S)))
cat(sprintf("V dimensions : %d x %d\n\n", nrow(res1$V), ncol(res1$V)))

# Relative reconstruction error of the rank-M approximation:
# ||X_M - X||_F / ||X||_F  (large here: a Gaussian matrix has a flat
# spectrum, so 10 of 80 components capture little energy).
X1_approx <- res1$U %*% diag(res1$S) %*% t(res1$V)
rel_error1 <- norm(X1_approx - X1, "F") / norm(X1, "F")
cat("Relative reconstruction error ||X_M - X||_F / ||X||_F :",
    format(rel_error1, scientific = TRUE), "\n")

cat(sprintf("\nTop %d singular values:\n", M1))
for (k in seq_len(M1)) {
  cat(sprintf("  s%d = %.4f\n", k, res1$S[k]))
}

# Didactic checks: singular values are sorted decreasingly, and at full rank
# (M = p) the truncated SVD reproduces X exactly up to floating-point error.
stopifnot(all(diff(res1$S) <= 1e-12))

res1_full <- solver$compute(X1, p1)
X1_full   <- res1_full$U %*% diag(res1_full$S) %*% t(res1_full$V)
rel_error_full <- norm(X1_full - X1, "F") / norm(X1, "F")
cat("\nFull-rank (M = p) relative reconstruction error :",
    format(rel_error_full, scientific = TRUE), "\n")
stopifnot(rel_error_full < 1e-10)
cat("Check passed: full-rank reconstruction is exact.\n")

# ==============================================================================
# Part 2: Gram / Kernel Path (p > 2*n)
# ==============================================================================
# For wide matrices the solver works on the n x n Gram matrix X X^T instead
# of X itself, which is much cheaper when p >> n.
# ==============================================================================

cat("\n", strrep("-", 70), "\n")
cat("Part 2: Gram Path (p > 2*n)\n")
cat(strrep("-", 70), "\n\n")

set.seed(7)
n2 <- 100L
p2 <- 500L
M2 <- 8L

cat(sprintf("Matrix size : %d x %d  (p > 2*n -> Gram path)\n", n2, p2))
cat(sprintf("M           : %d\n\n", M2))

X2 <- matrix(rnorm(n2 * p2), nrow = n2, ncol = p2)

res2 <- solver$compute(X2, M2)

cat(sprintf("U dimensions : %d x %d\n", nrow(res2$U), ncol(res2$U)))
cat(sprintf("S length     : %d\n", length(res2$S)))
cat(sprintf("V dimensions : %d x %d\n\n", nrow(res2$V), ncol(res2$V)))

X2_approx <- res2$U %*% diag(res2$S) %*% t(res2$V)
rel_error2 <- norm(X2_approx - X2, "F") / norm(X2, "F")
cat("Relative reconstruction error ||X_M - X||_F / ||X||_F :",
    format(rel_error2, scientific = TRUE), "\n")

# ==============================================================================
# Part 3: Orthogonality Check
# ==============================================================================
# The singular vector blocks must be orthonormal: U^T U = I and V^T V = I.
# ==============================================================================

cat("\n", strrep("-", 70), "\n")
cat("Part 3: Orthogonality Check\n")
cat(strrep("-", 70), "\n\n")

set.seed(99)
n3 <- 200L
p3 <- 150L
M3 <- 20L

cat(sprintf("Matrix size : %d x %d\n", n3, p3))
cat(sprintf("M           : %d\n\n", M3))

X3 <- matrix(rnorm(n3 * p3), nrow = n3, ncol = p3)

res3 <- solver$compute(X3, M3)

I_M       <- diag(M3)
UtU_error <- norm(crossprod(res3$U) - I_M, "F")
VtV_error <- norm(crossprod(res3$V) - I_M, "F")

cat("||U^T U - I||_F :", format(UtU_error, scientific = TRUE), "\n")
cat("||V^T V - I||_F :", format(VtV_error, scientific = TRUE), "\n")

stopifnot(UtU_error < 1e-10, VtV_error < 1e-10)
cat("\nCheck passed: U and V are orthonormal.\n")

cat("\n", strrep("=", 70), "\n")
cat("demo_svd_01_in_memory.R complete.\n")
